propagator/propagator.go

```diff
@@ -0,0 +1,144 @@
+// https://cloud.google.com/trace/docs/setup#force-trace
+const TraceContextHeaderName = "x-cloud-trace-context"
+
+// traceContextHeaderFormat is the regular expression pattan for valid Cloud Trace header value
+const traceContextHeaderFormat = "^(?P<trace_id>[0-9a-f]{32})/(?P<span_id>[0-9]{1,20})(;o=(?P<trace_flags>[0-9]))?$"
+
+// traceContextHeaderRe is a regular expression object of TraceContextHeaderFormat.
+var traceContextHeaderRe = regexp.MustCompile(traceContextHeaderFormat)
+
+var traceContextHeaders = []string{TraceContextHeaderName}
+
+type CloudTraceFormatPropagator struct{}
+
+func New() propagation.TextMapPropagator {
+	return CloudTraceFormatPropagator{}
+}
+
+func (p CloudTraceFormatPropagator) getHeaderValue(carrier propagation.TextMapCarrier) string {
+	return carrier.Get(TraceContextHeaderName)
+}
+
+// Inject injects a context to the carrier following Google Cloud Trace format.
+func (p CloudTraceFormatPropagator) Inject(ctx context.Context, carrier propagation.TextMapCarrier) {
+	span := trace.SpanFromContext(ctx)
+	sc := span.SpanContext()
+
+	// https://cloud.google.com/trace/docs/setup#force-trace
+	// Trace ID: 32-char hexadecimal value representing a 128-bit number.
+	// Span ID: decimal representation of the unsigned interger.
+	sidHex := sc.SpanID().String()
+	sid, err := strconv.ParseUint(sidHex, 16, 64)
+	if err != nil {
+		return
+	}
+	flag, err := strconv.Atoi(sc.TraceFlags().String())
+	if err != nil {
+		return
+	}
+	header := fmt.Sprintf("%s/%d;o=%d",
+		sc.TraceID().String(),
+		sid,
+		flag,
+	)
+	carrier.Set(TraceContextHeaderName, header)
+}
+
+// Extract extacts a context from the carrier if the header contains Google Cloud Trace header format.
+func (p CloudTraceFormatPropagator) Extract(ctx context.Context, carrier propagation.TextMapCarrier) context.Context {
+	header := p.getHeaderValue(carrier)
+	if header == "" {
+		return ctx
+	}
+
+	match := traceContextHeaderRe.FindStringSubmatch(header)
+	if match == nil {
+		return ctx
+	}
+	names := traceContextHeaderRe.SubexpNames()
+	var traceID, spanID, traceFlags string
+	for i, n := range names {
+		switch n {
+		case "trace_id":
+			traceID = match[i]
+		case "span_id":
+			spanID = match[i]
+		case "trace_flags":
+			traceFlags = match[i]
+		}
+	}
+	// non-recording Span
+	if traceID == strings.Repeat("0", 32) || spanID == "0" {
+		return ctx
+	}
+
+	// https://cloud.google.com/trace/docs/setup#force-trace
+	// Trace ID: 32-char hexadecimal value representing a 128-bit number.
+	// Span ID: decimal representation of the unsigned interger.
+	tid, err := trace.TraceIDFromHex(traceID)
+	if err != nil {
+		log.Printf("CloudTraceFormatPropagator: invalid trace id %#v: %v", traceID, err)
+		return ctx
+	}
+	sidUint, err := strconv.ParseUint(spanID, 10, 64)
+	if err != nil {
+		log.Printf("CloudTraceFormatPropagator: on span ID conversion: %v", err)
+		return ctx
+	}
+	sid, err := trace.SpanIDFromHex(fmt.Sprintf("%016x", sidUint))
```

Comment:
Same comment as above about the hex conversion, would using `binary.BigEndian.PutUint64()` or something similar work?